File: apps/app-frontend/src/helpers/theming.ts
import { invoke } from '@tauri-apps/api/core'

export interface ThemeFile {
	schemaVersion: number
	name: string
	colorTheme?: string
	accentColor?: string
	variables?: Record<string, string>
	css?: string
	font?: string
}

export interface InstalledTheme {
	id: string
	fileName: string
	theme: ThemeFile
}

export interface CssThemeError {
	message: string
}

// List all valid custom theme files in the configured themes folder
export async function listInstalledThemes(): Promise<InstalledTheme[]> {
	return (await invoke('plugin:theming|theming_list_installed_themes')) as InstalledTheme[]
}

// Validate an externally-picked theme file and copy it into the themes folder
export async function uploadTheme(sourcePath: string): Promise<InstalledTheme> {
	return (await invoke('plugin:theming|theming_upload_theme', {
		sourcePath,
	})) as InstalledTheme
}

// (Re-)register the configured themes folder with the file watcher so live
// edits hot-reload. Call after changing the themes folder path.
export async function watchThemesDir(): Promise<void> {
	await invoke('plugin:theming|theming_watch_dir')
}

// Open the native WebView inspector for debugging themes.
export async function openDevtools(): Promise<void> {
	await invoke('open_devtools')
}

// Parse theme CSS with css-tree in tolerant mode, collecting every syntax
// error so theme creators learn why a rule didn't apply (browsers silently
// drop invalid CSS with no feedback).
export async function validateThemeCss(css: string): Promise<CssThemeError[]> {
	const { parse } = await import('css-tree')
	const errors: CssThemeError[] = []
	parse(css, {
		positions: true,
		onParseError: (error) => {
			errors.push({ message: error.formattedMessage || error.message })
		},
	})
	return errors
}

File: apps/app/src/api/theming.rs
use crate::api::Result;
use theseus::prelude::*;

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("theming")
        .invoke_handler(tauri::generate_handler![
            theming_list_installed_themes,
            theming_upload_theme,
            theming_watch_dir,
        ])
        .build()
}

// List all valid custom theme files in the configured themes folder
// invoke('plugin:theming|theming_list_installed_themes')
#[tauri::command]
pub async fn theming_list_installed_themes()
-> Result<Vec<theming::InstalledTheme>> {
    let res = theming::list_installed_themes().await?;
    Ok(res)
}

// Validate an externally-picked theme file and copy it into the themes folder
// invoke('plugin:theming|theming_upload_theme', { sourcePath })
#[tauri::command]
pub async fn theming_upload_theme(
    source_path: String,
) -> Result<theming::InstalledTheme> {
    let res = theming::upload_theme(&source_path).await?;
    Ok(res)
}

// (Re-)register the configured themes folder with the file watcher so live
// edits hot-reload. Called after the themes folder setting changes.
// invoke('plugin:theming|theming_watch_dir')
#[tauri::command]
pub async fn theming_watch_dir() -> Result<()> {
    theming::refresh_themes_watch().await?;
    Ok(())
}
